### PR TITLE
slices for 'tuple' properties + detailed exclusion and invalid particle error messages

### LINKED ISSUE #1214: Interface exception message "Set particle position first." misleading
I did not check all of them, but some of the errors that say _"Set particle position first"_ thrown on failed particle property changes have nothing to do with the positions. We should run through the `set_particle..() `methods in particle_data.cpp and give meaningful error messages. Most of them are _"Particle with ID xyz does not exists"_ I think.

### LINKED ISSUE #1563: Exclusions and bonds don't work on ParticleSlice
Because they can have variable length, so they can not be cast into a numpy array. If this can't be fixed, they have to be removed from the slice.

```python
from __future__ import print_function
import espressomd as es

s = es.System()
s.part.add(pos=[0,0,0])
s.part.add(pos=[0,0,0], exclusions=[0])
s.part.add(pos=[0,0,0])

print(s.part[:].exclusions)
```

### PR DESCRIPTION
**Pull request by KonradBreitsprecher**
Fixes #1214, fixes #1563 

- Extended exclusion error message for adding/deleting non-valid ids
- delete_exclusion now accepts lists and single int
- Checks if attribute is "exclusions", "bonds", "vs_relative", "swimming" and returns a tuple of tuples.

### CODE REVIEW

**Comment by fweik** [comment]
Does the setting of bonds/exclusions from a list work too? Also this should be covered in the particle_slice test.

**Comment by KonradBreitsprecher** [comment]
I will check..

**Comment by KonradBreitsprecher** [comment]
Setting slices now works for bonds, exclusions and vs_relative.
- Extensive tests in particle_slice.py
- Brief docu in part.rst
Please review @KaiSzuttor .

File: src/python/espressomd/particle_data.pyx [hunk 1]
```diff
@@ -1019,24 +1024,31 @@ cdef class ParticleHandle(object):
 
             def __set__(self, _rot):
                 cdef int rot
-                check_type_or_throw_except(_rot,3,int,"The rotation flag is given as a tuple of 3 integes.")
-                    
-                rot=0
-                if _rot[0]: rot+=ROT_X
-                if _rot[1]: rot+=ROT_Y
-                if _rot[2]: rot+=ROT_Z
+                check_type_or_throw_except(
```
Comment (KaiSzuttor) [inline src/python/espressomd/particle_data.pyx:1030]: has to be given as

File: src/python/espressomd/particle_data.pyx [hunk 2]
```diff
@@ -1285,6 +1315,10 @@ cdef class ParticleHandle(object):
         bond_info[0] = bond[0]._bond_id
         for i in range(1, len(bond)):
             bond_info[i] = bond[i]
+        if self.id in bond[1:]:
+            raise Exception(
+                "Bond partners {} includes the particle {} itself.".format(bond[1:], self.id))
```
Comment (KaiSzuttor) [inline src/python/espressomd/particle_data.pyx]: include~~s~~

File: src/python/espressomd/particle_data.pyx [hunk 3]
```diff
@@ -1534,26 +1574,13 @@ cdef class _ParticleSliceImpl(object):
...
-                raise TypeError("Unexpected exclusion partner type.")
+        def add_exclusion(self, _partner):
+            for i in range(len(self.id_selection)):
+                ParticleHandle(self.id_selection[i]).add_exclusion(_partner)
 
-        def delete_exclusions(self):
-            for _id in self.id_selection:
-                ParticleHandle(_id).delete_exclusions()
+        def delete_exclusion(self, _partner):
+            for i in range(len(self.id_selection)):
```
Comment (KaiSzuttor) [inline src/python/espressomd/particle_data.pyx]:
```python
for i in self.id_selection:
    ParticleHandle(i).delete_exclusion(_partner)
```

File: src/python/espressomd/particle_data.pyx [hunk 4]
```diff
@@ -1578,25 +1605,17 @@ cdef class _ParticleSliceImpl(object):
 
         """
 
-        bond = list(_bond)  # As we will modify it
         for i in range(len(self.id_selection)):
```
Comment (KaiSzuttor) [inline src/python/espressomd/particle_data.pyx]:
```python
for i in self.id_selection:
    ...
```

File: src/python/espressomd/particle_data.pyx [hunk 5]
```diff
@@ -1578,25 +1605,17 @@ cdef class _ParticleSliceImpl(object):
...
+            ParticleHandle(self.id_selection[i]).add_bond(_bond)
 
     def delete_bond(self, _bond):
         """
         Delete a single bond from the particles.
 
         """
 
-        bond = list(_bond)  # as we modify it
         for i in range(len(self.id_selection)):
```
Comment (KaiSzuttor) [inline src/python/espressomd/particle_data.pyx]:
```python
for i in...
```

File: src/python/espressomd/particle_data.pyx [hunk 6]
```diff
@@ -1839,18 +1857,18 @@ cdef class ParticleList(object):
         """
         Write the positions and velocities of particles with specified
         types to a VTK file.
- 
+
         Parameters
         ----------
- 
+
```
Comment (KaiSzuttor) [inline src/python/espressomd/particle_data.pyx]: remote blank line

File: src/python/espressomd/particle_data.pyx [hunk 7]
```diff
@@ -1839,18 +1857,18 @@ cdef class ParticleList(object):
...
         fname: :obj:`str`
```
Comment (KaiSzuttor) [inline src/python/espressomd/particle_data.pyx]: fname_ : :obj:`str`
blank between parameter name and colon for all parameters

File: src/python/espressomd/particle_data.pyx [hunk 8]
```diff
@@ -1958,48 +1987,82 @@ def _add_particle_slice_properties():
         Setter function that sets attribute on every member of particle_slice.
         If values contains only one element, all members are set to it. If it
         contains as many elements as there are members, each of them gets set
-        to the corresponding one.
-        
+        to the corresponding one. For attributes that are lists of various lenght,
+        (bonds, exclusions) the nesting level decides if it is one-for-all or one-for-each.
+
+
```
Comment (KaiSzuttor) [inline src/python/espressomd/particle_data.pyx]: remove one blank

File: src/python/espressomd/utils.pyx [hunk 9]
```diff
@@ -226,4 +226,17 @@ def get_unravelled_index(len_dims, n_dims, flattened_index):
     for i in range(n_dims):
         out[i] = unravelled_index_out[i]
     return out
-    
+   
+def nest_level(obj):
+
```
Comment (KaiSzuttor) [inline src/python/espressomd/utils.pyx]: please shortly describe what the function does

File: testsuite/particle_slice.py [hunk 10]
```diff
@@ -9,6 +9,7 @@ class ParticleSliceTest(ut.TestCase):
 
     state = [[0, 0, 0], [0, 0, 1]]
```
Comment (KaiSzuttor) [inline testsuite/particle_slice.py:10]: autopep8 -i on that file

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

**Comment by fweik** [comment]
/builds/espressomd/espresso/src/python/espressomd/utils.pyx:243:35: undeclared name not builtin: nest_level

**Comment by fweik** [comment]
Not sure if I messed up the merge or if this stuff was broken before, feel free to reset to a8e6367.

**Review by KaiSzuttor — APPROVED** [review]

**Review by fweik — APPROVED** [review]

**Pull request merged**